### PR TITLE
add Optional typescript module

### LINKED ISSUE #4057: Add Typescript Optional module
following discussion with @hdurix
The idea is to add the possibility to add the Optional utility to typescript domain on client app

### PR DESCRIPTION
**Pull request by thomaslmbrt**
Close #4057

### CODE REVIEW

File: src/main/java/tech/jhipster/lite/generator/typescript/optional/domain/OptionalTypescriptModuleFactory.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+public class OptionalTypescriptModuleFactory {
+
+  public JHipsterModule buildModule(JHipsterModuleProperties properties) {
+    Assert.notNull("properties", properties);
+
+    //@formatter:off
+    return moduleBuilder(properties)
+      .files()
+      .add(from("typescript/webapp/common/domain/optional/").file("Optional.ts"), to("src/webapp/common/domain/Optional.ts"))
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/typescript/optional/domain/OptionalTypescriptModuleFactory.java:19]:
```suggestion
      .add(from("typescript/webapp/common/domain/optional/").file("Optional.ts"), to("src/webapp/app/common/domain/Optional.ts"))
```

File: src/test/features/optional-typescript.feature [hunk 2]
```diff
@@ -0,0 +1,6 @@
+Feature: Optional-typescript module
+
+  Scenario: Add Optional class domain
+    When I apply "optional-typescript" module to default project without parameters
+    Then I should have files in "/src/webapp/common/domain/"
```
Comment (pascalgrimaud) [inline src/test/features/optional-typescript.feature:5]:
```suggestion
    Then I should have files in "/src/webapp/app/common/domain/"
```

File: src/test/java/tech/jhipster/lite/generator/typescript/domain/optional/OptionalTypescriptModuleFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import org.junit.jupiter.api.Test;
+import tech.jhipster.lite.TestFileUtils;
+import tech.jhipster.lite.UnitTest;
+import tech.jhipster.lite.generator.typescript.optional.domain.OptionalTypescriptModuleFactory;
+import tech.jhipster.lite.module.domain.JHipsterModule;
+import tech.jhipster.lite.module.domain.JHipsterModulesFixture;
+import tech.jhipster.lite.module.domain.properties.JHipsterModuleProperties;
+
+@UnitTest
+public class OptionalTypescriptModuleFactoryTest {
```
Comment (pascalgrimaud) [inline src/test/java/tech/jhipster/lite/generator/typescript/domain/optional/OptionalTypescriptModuleFactoryTest.java:14]:
```suggestion
class OptionalTypescriptModuleFactoryTest {
```

File: src/test/java/tech/jhipster/lite/generator/typescript/domain/optional/OptionalTypescriptModuleFactoryTest.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  private static final OptionalTypescriptModuleFactory factory = new OptionalTypescriptModuleFactory();
+
+  @Test
+  void shouldCreateOptionalTypescriptModule() {
+    JHipsterModuleProperties properties = JHipsterModulesFixture.propertiesBuilder(TestFileUtils.tmpDirForTest()).build();
+
+    JHipsterModule module = factory.buildModule(properties);
+
+    assertThatModuleWithFiles(module)
+      .hasFile("src/webapp/common/domain/Optional.ts")
```
Comment (pascalgrimaud) [inline src/test/java/tech/jhipster/lite/generator/typescript/domain/optional/OptionalTypescriptModuleFactoryTest.java:25]:
```suggestion
      .hasFile("src/webapp/app/common/domain/Optional.ts")
```

File: src/main/java/tech/jhipster/lite/generator/typescript/optional/primary/OptionalTypescriptModuleConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import static tech.jhipster.lite.generator.JHLiteModuleSlug.TYPESCRIPT;
+
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import tech.jhipster.lite.generator.typescript.optional.application.OptionalTypescriptApplicationService;
+import tech.jhipster.lite.module.domain.resource.JHipsterModuleOrganization;
+import tech.jhipster.lite.module.domain.resource.JHipsterModuleResource;
+
+@Configuration
+public class OptionalTypescriptModuleConfiguration {
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/typescript/optional/primary/OptionalTypescriptModuleConfiguration.java:13]:
```suggestion
class OptionalTypescriptModuleConfiguration {
```

**Review by pascalgrimaud — CHANGES_REQUESTED** [review]

File: src/main/resources/generator/typescript/webapp/common/domain/optional/Optional.ts [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+  static ofUndefinable<Value>(value: Value | undefined): Optional<Value> {
+    if (value === undefined) {
+      return Optional.empty();
+    }
+
+    return Optional.of(value);
+  }
+
+  abstract map<Output>(mapper: (value: Value) => Output): Optional<Output>;
```
Comment (DamnClin) [inline src/main/resources/generator/typescript/webapp/common/domain/optional/Optional.ts:18]: I think `or(() => Optional<Value>)` should be added

**Comment by pascalgrimaud** [comment]
Another comment:

image

Can you put a different `package-info.java` inside each folder plz (instead at the root of typescript folder): 
- core
- optional

File: src/main/resources/generator/typescript/webapp/common/domain/optional/Optional.ts (hunk 6 above)
Comment (pascalgrimaud) [reply src/main/resources/generator/typescript/webapp/common/domain/optional/Optional.ts:18]: this part can be improved in another PR, and for JHLite + Generated part too
Let's merge this for now

Comment (pascalgrimaud) [reply src/main/resources/generator/typescript/webapp/common/domain/optional/Optional.ts:18]: oh forget my comment, it was already added ! :)

**Comment by pascalgrimaud** [comment]
Thanks a lot for your contribution @thomaslmbrt !

**Pull request merged**